File: src/emuchef_editor/core/refs/ref_index.py
"""Reference index helpers for authored recipe documents."""

from __future__ import annotations

from dataclasses import dataclass

from emuchef.domain import InputType, Recipe, RuntimeValueType
from emuchef.planner.contracts import RUNTIME_ARTIFACT_FIELDS
from emuchef.steps import builtin_step_registry


@dataclass(frozen=True, slots=True)
class RefCandidate:
    """A typed explicit authored ref that can be offered in structured pickers."""

    ref: str
    label: str
    value_type: RuntimeValueType
    source_kind: str
    source_id: str


@dataclass(frozen=True, slots=True)
class RefIndex:
    input_refs: tuple[str, ...]
    artifact_refs: tuple[str, ...]
    step_refs: tuple[str, ...]
    step_output_refs: tuple[str, ...]
    candidates: tuple[RefCandidate, ...]

    @property
    def all_refs(self) -> tuple[str, ...]:
        return self.input_refs + self.artifact_refs + self.step_refs + self.step_output_refs


def build_ref_index(recipe: Recipe) -> RefIndex:
    input_refs = tuple(f"inputs.{input_id}" for input_id in sorted(recipe.inputs))
    artifact_refs = tuple(
        f"artifacts.{artifact_id}.{field}"
        for artifact_id in sorted(recipe.artifacts)
        for field in sorted(RUNTIME_ARTIFACT_FIELDS)
    )
    step_refs = tuple(f"steps.{step.id}" for step in recipe.steps)
    step_output_refs = tuple(
        f"steps.{step.id}.outputs.{output_name}"
        for step in recipe.steps
        for output_name in _step_output_names(step.type)
    )
    candidates = (
        _input_candidates(recipe)
        + _artifact_candidates(recipe)
        + _step_output_candidates(recipe)
    )
    return RefIndex(
        input_refs=input_refs,
        artifact_refs=artifact_refs,
        step_refs=step_refs,
        step_output_refs=step_output_refs,
        candidates=candidates,
    )


def _step_output_names(step_type) -> tuple[str, ...]:
    plugin = builtin_step_registry().get(step_type)
    if plugin is None:
        return ()
    return tuple(output.name for output in plugin.outputs)


def _input_candidates(recipe: Recipe) -> tuple[RefCandidate, ...]:
    candidates: list[RefCandidate] = []
    for input_id, declaration in recipe.inputs.items():
        candidates.append(
            RefCandidate(
                ref=f"inputs.{input_id}",
                label=f"Input · {input_id}",
                value_type=_input_value_type(declaration.type, declaration.multiple),
                source_kind="input",
                source_id=input_id,
            )
        )
    return tuple(candidates)


def _artifact_candidates(recipe: Recipe) -> tuple[RefCandidate, ...]:
    candidates: list[RefCandidate] = []
    for artifact_id in recipe.artifacts:
        for field in sorted(RUNTIME_ARTIFACT_FIELDS):
            value_type = _ARTIFACT_FIELD_TYPES.get(field)
            if value_type is None:
                continue
            candidates.append(
                RefCandidate(
                    ref=f"artifacts.{artifact_id}.{field}",
                    label=f"Artifact · {artifact_id}.{field}",
                    value_type=value_type,
                    source_kind="artifact",
                    source_id=artifact_id,
                )
            )
    return tuple(candidates)


def _step_output_candidates(recipe: Recipe) -> tuple[RefCandidate, ...]:
    candidates: list[RefCandidate] = []
    for step in recipe.steps:
        for output_name in _step_output_names(step.type):
            value_type = _PRIMARY_OUTPUT_TYPES.get((step.type, output_name))
            if value_type is None:
                continue
            candidates.append(
                RefCandidate(
                    ref=f"steps.{step.id}.outputs.{output_name}",
                    label=f"Step Output · {step.id}.{output_name}",
                    value_type=value_type,
                    source_kind="step_output",
                    source_id=step.id,
                )
            )
    return tuple(candidates)


def _input_value_type(input_type: InputType, multiple: bool) -> RuntimeValueType:
    if multiple:
        return RuntimeValueType.PATH_LIST
    if input_type is InputType.DIRECTORY:
        return RuntimeValueType.DIRECTORY_PATH
    return RuntimeValueType.FILE_PATH


_ARTIFACT_FIELD_TYPES: dict[str, RuntimeValueType] = {
    "status": RuntimeValueType.STRING,
    "local_path": RuntimeValueType.FILE_PATH,
    "resolved_url": RuntimeValueType.STRING,
    "filename": RuntimeValueType.STRING,
    "cache_hit": RuntimeValueType.BOOLEAN,
    "error": RuntimeValueType.STRING,
}

_PRIMARY_OUTPUT_TYPES: dict[tuple[object, str], RuntimeValueType] = {
    (plugin.type, output.name): output.value_type
    for plugin in builtin_step_registry().plugins
    for output in plugin.outputs
}

File: crates/emuchef-rust-backend/src/ref_index.rs
//! RefIndex generation for the Phase 6H modeled authored recipe surface.
//!
//! The Python backend remains the reference implementation. This module mirrors
//! its current `build_ref_index` behavior for recipe inputs, runtime artifact
//! fields, authored step ids, and declared StepSpec outputs. It intentionally
//! does not inspect authored param refs, artifact groups, planner state,
//! catalog context, or executor/device data.

use serde_json::{json, Value};

use crate::model::Recipe;
use crate::step_specs;

const RUNTIME_ARTIFACT_FIELDS: &[&str] = &[
    "cache_hit",
    "error",
    "filename",
    "local_path",
    "resolved_url",
    "status",
];

pub fn ref_index_to_dto(recipe: &Recipe) -> Value {
    let input_refs = input_refs(recipe);
    let artifact_refs = artifact_refs(recipe);
    let step_refs = step_refs(recipe);
    let step_output_refs = step_output_refs(recipe);
    let all_refs = input_refs
        .iter()
        .chain(artifact_refs.iter())
        .chain(step_refs.iter())
        .chain(step_output_refs.iter())
        .cloned()
        .collect::<Vec<_>>();
    let candidates = input_candidates(recipe)
        .into_iter()
        .chain(artifact_candidates(recipe))
        .chain(step_output_candidates(recipe))
        .collect::<Vec<_>>();

    json!({
        "inputRefs": input_refs,
        "artifactRefs": artifact_refs,
        "stepRefs": step_refs,
        "stepOutputRefs": step_output_refs,
        "allRefs": all_refs,
        "candidates": candidates,
    })
}

fn input_refs(recipe: &Recipe) -> Vec<String> {
    let mut refs = recipe
        .inputs
        .keys()
        .map(|input_id| format!("inputs.{input_id}"))
        .collect::<Vec<_>>();
    refs.sort();
    refs
}

fn artifact_refs(recipe: &Recipe) -> Vec<String> {
    let mut artifact_ids = recipe.artifacts.keys().collect::<Vec<_>>();
    artifact_ids.sort();
    artifact_ids
        .into_iter()
        .flat_map(|artifact_id| {
            RUNTIME_ARTIFACT_FIELDS
                .iter()
                .map(move |field| format!("artifacts.{artifact_id}.{field}"))
        })
        .collect()
}

fn step_refs(recipe: &Recipe) -> Vec<String> {
    recipe
        .steps
        .iter()
        .map(|step| format!("steps.{}", step.id))
        .collect()
}

fn step_output_refs(recipe: &Recipe) -> Vec<String> {
    recipe
        .steps
        .iter()
        .flat_map(|step| {
            step_specs::step_spec_for(&step.type_name)
                .into_iter()
                .flat_map(move |spec| {
                    spec.outputs
                        .into_iter()
                        .map(move |output| format!("steps.{}.outputs.{}", step.id, output.name))
                })
        })
        .collect()
}

fn input_candidates(recipe: &Recipe) -> Vec<Value> {
    recipe
        .inputs
        .iter()
        .map(|(input_id, input)| {
            json!({
                "ref": format!("inputs.{input_id}"),
                "label": format!("Input \u{00b7} {input_id}"),
                "valueType": input_value_type(&input.type_name, input.multiple),
                "sourceKind": "input",
                "sourceId": input_id,
            })
        })
        .collect()
}

fn artifact_candidates(recipe: &Recipe) -> Vec<Value> {
    recipe
        .artifacts
        .keys()
        .flat_map(|artifact_id| {
            RUNTIME_ARTIFACT_FIELDS.iter().map(move |field| {
                json!({
                    "ref": format!("artifacts.{artifact_id}.{field}"),
                    "label": format!("Artifact \u{00b7} {artifact_id}.{field}"),
                    "valueType": artifact_field_value_type(field),
                    "sourceKind": "artifact",
                    "sourceId": artifact_id,
                })
            })
        })
        .collect()
}

fn step_output_candidates(recipe: &Recipe) -> Vec<Value> {
    recipe
        .steps
        .iter()
        .flat_map(|step| {
            step_specs::step_spec_for(&step.type_name)
                .into_iter()
                .flat_map(move |spec| {
                    spec.outputs.into_iter().map(move |output| {
                        json!({
                            "ref": format!("steps.{}.outputs.{}", step.id, output.name),
                            "label": format!("Step Output \u{00b7} {}.{}", step.id, output.name),
                            "valueType": output.value_type,
                            "sourceKind": "step_output",
                            "sourceId": step.id,
                        })
                    })
                })
        })
        .collect()
}

fn input_value_type(type_name: &str, multiple: bool) -> &'static str {
    if multiple {
        "path_list"
    } else if type_name == "directory" {
        "directory_path"
    } else {
        "file_path"
    }
}

fn artifact_field_value_type(field: &str) -> &'static str {
    match field {
        "cache_hit" => "boolean",
        "local_path" => "file_path",
        "error" | "filename" | "resolved_url" | "status" => "string",
        _ => unreachable!("runtime artifact fields are fixed by RUNTIME_ARTIFACT_FIELDS"),
    }
}
